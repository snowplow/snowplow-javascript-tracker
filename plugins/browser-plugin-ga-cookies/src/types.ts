export interface UACookies {
  __utma?: string;
  __utmb?: string;
  __utmc?: string;
  __utmv?: string;
  __utmz?: string;
  _ga?: string;
  [key: string]: string | undefined;
}

export interface GA4Cookies {
  _ga?: string;
  session_cookies?: GA4SessionCookies[];
  [key: string]: unknown;
}

export interface GA4SessionCookies {
  measurement_id: string;
  session_cookie?: string;
}
