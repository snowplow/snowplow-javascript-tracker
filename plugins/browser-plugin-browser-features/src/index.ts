/*
 * Copyright (c) 2022 Snowplow Analytics Ltd, 2010 Anthon Pang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { isFunction, BrowserPlugin, BrowserTracker } from '@snowplow/browser-tracker-core';

declare global {
  interface MimeTypeArray {
    [index: string]: MimeType;
  }
}

const pluginMap: Record<string, string> = {
  // document types
  pdf: 'application/pdf',

  // media players
  qt: 'video/quicktime',
  realp: 'audio/x-pn-realaudio-plugin',
  wma: 'application/x-mplayer2',

  // interactive multimedia
  dir: 'application/x-director',
  fla: 'application/x-shockwave-flash',

  // RIA
  java: 'application/x-java-vm',
  gears: 'application/x-googlegears',
  ag: 'application/x-silverlight',
};

/**
 * Adds the available MIME Types to each event
 */
export function BrowserFeaturesPlugin(): BrowserPlugin {
  return {
    activateBrowserPlugin: (tracker: BrowserTracker) => {
      const navigatorAlias = navigator;
      // General plugin detection
      if (navigatorAlias.mimeTypes && navigatorAlias.mimeTypes.length) {
        for (const i in pluginMap) {
          if (Object.prototype.hasOwnProperty.call(pluginMap, i)) {
            let mimeType = navigatorAlias.mimeTypes[pluginMap[i]];
            if (mimeType) {
              tracker.core.addPayloadPair('f_' + i, mimeType.enabledPlugin ? '1' : '0');
            }
          }
        }
      }

      // Firefox
      if (isFunction((window as any).GearsFactory)) {
        tracker.core.addPayloadPair('f_gears', '1');
      }
    },
  };
}
