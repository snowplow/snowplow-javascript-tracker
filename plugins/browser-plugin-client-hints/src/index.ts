import { BrowserPlugin } from '@snowplow/browser-tracker-core';
import { HttpClientHints } from './contexts';

interface NavigatorUABrandVersion {
  brand: string; // "Google Chrome"
  version: string; // "84"
}

interface UADataValues {
  platform: string; // "PhoneOS"
  platformVersion: string; // "10A"
  architecture: string; // "arm"
  model: string; // "X644GTM"
  uaFullVersion: string; // "73.32.AGX.5"
}

declare global {
  interface Navigator {
    userAgentData: {
      mobile: boolean;
      brands: Array<NavigatorUABrandVersion>;
      getHighEntropyValues: (hints: Array<string>) => Promise<UADataValues>;
    };
  }
}

let uaClientHints: HttpClientHints;

/**
 * This function makes sure that the expected array is returned as an array instead of an object.
 * It handles a problem that in some cases the `navigator.userAgentData.brands` was returned as an object instead of array.
 */
function forceArray<T>(array: T[] | Record<string, T>): T[] {
  if (Array.isArray(array)) return array;

  if (Object.prototype.toString.call(array) === '[object Object]') {
    return Object.keys(array).map((e) => {
      return array[e];
    });
  }

  return [];
}

/**
 * Returns the client-hints brands, ensuring no additional properties.
 */
function getBrands(brands: Array<NavigatorUABrandVersion>): Array<NavigatorUABrandVersion> {
  return brands.map((b) => {
    const { brand, version } = b;
    return { brand, version };
  });
}

/**
 * Validates whether userAgentData is compliant to the client-hints interface.
 * https://wicg.github.io/ua-client-hints/#interface
 */
function validClientHints(hints: HttpClientHints): boolean {
  if (!hints || typeof hints.isMobile !== 'boolean' || !Array.isArray(hints.brands)) {
    return false;
  }

  if (
    hints.brands.length === 0 ||
    hints.brands.some((brand) => typeof brand.brand !== 'string' || typeof brand.version !== 'string')
  ) {
    return false;
  }

  return true;
}

/**
 * Attaches Client Hint information where available
 * @param includeHighEntropy - Should high entropy values be included
 */
export function ClientHintsPlugin(includeHighEntropy?: boolean): BrowserPlugin {
  const populateClientHints = () => {
    const navigatorAlias = navigator;
    const uaData = navigatorAlias.userAgentData;

    if (uaData) {
      let candidateHints: HttpClientHints;
      candidateHints = {
        isMobile: uaData.mobile,
        brands: getBrands(forceArray(uaData.brands)),
      };
      if (includeHighEntropy && uaData.getHighEntropyValues) {
        uaData
          .getHighEntropyValues(['platform', 'platformVersion', 'architecture', 'model', 'uaFullVersion'])
          .then((res) => {
            candidateHints.architecture = res.architecture;
            candidateHints.model = res.model;
            candidateHints.platform = res.platform;
            candidateHints.uaFullVersion = res.uaFullVersion;
            candidateHints.platformVersion = res.platformVersion;
          });
      }

      if (validClientHints(candidateHints)) {
        uaClientHints = candidateHints;
      }
    }
  };

  return {
    activateBrowserPlugin: () => {
      if (!uaClientHints) {
        populateClientHints();
      }
    },
    contexts: () => {
      if (uaClientHints) {
        return [
          {
            schema: 'iglu:org.ietf/http_client_hints/jsonschema/1-0-0',
            data: uaClientHints,
          },
        ];
      }
      return [];
    },
  };
}
