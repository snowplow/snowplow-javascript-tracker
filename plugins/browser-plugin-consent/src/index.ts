/*
 * Copyright (c) 2022 Snowplow Analytics Ltd, 2010 Anthon Pang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { BrowserPlugin, BrowserTracker, dispatchToTrackersInCollection } from '@snowplow/browser-tracker-core';
import {
  buildConsentGranted,
  buildConsentWithdrawn,
  CommonEventProperties,
  ConsentGrantedEvent,
  ConsentWithdrawnEvent,
  Logger,
} from '@snowplow/tracker-core';
import { Gdpr } from './contexts';

export { ConsentGrantedEvent, ConsentWithdrawnEvent };

export enum gdprBasis {
  consent = 'consent',
  contract = 'contract',
  legalObligation = 'legal_obligation',
  vitalInterests = 'vital_interests',
  publicTask = 'public_task',
  legitimateInterests = 'legitimate_interests',
}

export type GdprBasis = keyof typeof gdprBasis;

/** The Configuration for the GDPR Context */
export interface GdprContextConfiguration {
  /** The basis for why the document will be processed */
  basisForProcessing: GdprBasis;
  /** An identifier for the document */
  documentId?: string;
  /** The version of the document */
  documentVersion?: string;
  /** A descrtiption of the document */
  documentDescription?: string;
}

const _trackers: Record<string, BrowserTracker> = {};
const _context: Record<string, Gdpr> = {};
let LOG: Logger;

/**
 * The Consent Plugin
 *
 * Adds Consent Granted and Withdrawn events
 * and the ability to add the GDPR context to events
 */
export function ConsentPlugin(): BrowserPlugin {
  let trackerId: string;

  return {
    activateBrowserPlugin: (tracker) => {
      trackerId = tracker.id;
      _trackers[tracker.id] = tracker;
    },
    contexts: () => {
      if (_context[trackerId]) {
        return [
          {
            schema: 'iglu:com.snowplowanalytics.snowplow/gdpr/jsonschema/1-0-0',
            data: _context[trackerId],
          },
        ];
      }

      return [];
    },
    logger: (logger) => {
      LOG = logger;
    },
  };
}

/**
 * Enable the GDPR context for each event
 * @param configuration - the configuration for the GDPR context
 * @param trackers - The tracker identifiers which should have the GDPR context enabled
 */
export function enableGdprContext(
  configuration: GdprContextConfiguration,
  trackers: Array<string> = Object.keys(_trackers)
) {
  const { basisForProcessing, documentId, documentVersion, documentDescription } = configuration;
  let basis = gdprBasis[basisForProcessing];

  if (!basis) {
    LOG.warn(
      'enableGdprContext: basisForProcessing must be one of: consent, contract, legalObligation, vitalInterests, publicTask, legitimateInterests'
    );
    return;
  } else {
    trackers.forEach((t) => {
      if (_trackers[t]) {
        _context[t] = {
          basisForProcessing: basis,
          documentId: documentId ?? null,
          documentVersion: documentVersion ?? null,
          documentDescription: documentDescription ?? null,
        };
      }
    });
  }
}

/**
 * Track a consent granted action
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function trackConsentGranted(
  event: ConsentGrantedEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
    const builtEvent = buildConsentGranted(event);
    t.core.track(
      builtEvent.event,
      event.context ? event.context.concat(builtEvent.context) : builtEvent.context,
      event.timestamp
    );
  });
}

/**
 * Track a consent withdrawn action
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function trackConsentWithdrawn(
  event: ConsentWithdrawnEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
    const builtEvent = buildConsentWithdrawn(event);
    t.core.track(
      builtEvent.event,
      event.context ? event.context.concat(builtEvent.context) : builtEvent.context,
      event.timestamp
    );
  });
}
