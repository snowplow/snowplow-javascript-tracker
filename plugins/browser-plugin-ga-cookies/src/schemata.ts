export const UNIVERSAL_ANALYTICS_COOKIES_SCHEMA = 'iglu:com.google.analytics/cookies/jsonschema/1-0-0';
export const GOOGLE_ANALYTICS_4_COOKIES_SCHEMA = 'iglu:com.google.ga4/cookies/jsonschema/1-0-0';
