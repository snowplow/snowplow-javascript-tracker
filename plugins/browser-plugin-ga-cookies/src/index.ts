import { SelfDescribingJson } from '@snowplow/tracker-core';
import { BrowserPlugin, cookie } from '@snowplow/browser-tracker-core';
import { GA4Cookies, GA4SessionCookies, UACookies } from './types';
import { GOOGLE_ANALYTICS_4_COOKIES_SCHEMA, UNIVERSAL_ANALYTICS_COOKIES_SCHEMA } from './schemata';

interface GACookiesPluginOptions {
  ua?: boolean;
  ga4?: boolean;
  ga4MeasurementId?: string | string[];
  cookiePrefix?: string | string[];
}

const defaultPluginOptions: GACookiesPluginOptions = {
  ua: true,
  ga4: false,
  ga4MeasurementId: '',
  cookiePrefix: [],
};

/**
 * Captures the GA cookies on a page and sends as context on each event
 * @param pluginOptions.ua - Send Universal Analytics specific cookie values.
 * @param pluginOptions.ga4 - Send Google Analytics 4 specific cookie values.
 * @param pluginOptions.ga4MeasurementId - Measurement id/ids to search the Google Analytics 4 session cookie. Can be a single measurement id as a string or an array of measurement id strings. The cookie has the form of _ga_<container-id> where <container-id> is the data stream container id.
 * @param pluginOptions.cookiePrefix - Cookie prefix set on the Google Analytics 4 cookies using the cookie_prefix option of the gtag.js tracker.
 */
export function GaCookiesPlugin(pluginOptions: GACookiesPluginOptions = defaultPluginOptions): BrowserPlugin {
  return {
    contexts: () => {
      const contexts: SelfDescribingJson<Record<string, unknown>>[] = [];
      const { ga4, ga4MeasurementId, ua, cookiePrefix } = { ...defaultPluginOptions, ...pluginOptions };
      const GA_USER_COOKIE = '_ga';
      const GA4_MEASUREMENT_ID_PREFIX = 'G-';
      const GA4_COOKIE_PREFIX = '_ga_';

      if (ua) {
        const uaCookiesContext: SelfDescribingJson<UACookies> = {
          schema: UNIVERSAL_ANALYTICS_COOKIES_SCHEMA,
          data: {},
        };
        ['__utma', '__utmb', '__utmc', '__utmv', '__utmz', GA_USER_COOKIE].forEach(function (cookieType) {
          var value = cookie(cookieType);
          if (value) {
            uaCookiesContext.data[cookieType] = value;
          }
        });
        contexts.push(uaCookiesContext);
      }

      if (ga4) {
        const cookiePrefixes = Array.isArray(cookiePrefix) ? [...cookiePrefix] : [cookiePrefix];

        /* We also will search for the default (no prefix) in every case. */
        cookiePrefixes.unshift('');
        cookiePrefixes.forEach((prefix) => {
          const userCookie = cookie(prefix + GA_USER_COOKIE);
          const sessionCookies: GA4SessionCookies[] = [];
          if (ga4MeasurementId) {
            const measurementIdentifiers = Array.isArray(ga4MeasurementId) ? [...ga4MeasurementId] : [ga4MeasurementId];

            measurementIdentifiers.forEach(function (cookieIdentifier) {
              const sessionCookieValue = cookie(
                prefix + cookieIdentifier.replace(GA4_MEASUREMENT_ID_PREFIX, GA4_COOKIE_PREFIX)
              );
              if (sessionCookieValue) {
                sessionCookies.push({
                  measurement_id: cookieIdentifier,
                  session_cookie: sessionCookieValue,
                });
              }
            });
          }

          if (!userCookie && !sessionCookies.length) {
            return;
          }

          const ga4CookiesContext: SelfDescribingJson<GA4Cookies> = {
            schema: GOOGLE_ANALYTICS_4_COOKIES_SCHEMA,
            data: {
              _ga: userCookie,
              session_cookies: sessionCookies.length ? sessionCookies : undefined,
              cookie_prefix: prefix || undefined,
            },
          };

          contexts.push(ga4CookiesContext);
        });
      }

      return contexts;
    },
  };
}
