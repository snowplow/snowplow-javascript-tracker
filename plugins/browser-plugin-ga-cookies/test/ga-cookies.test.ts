import { buildLinkClick, trackerCore } from '@snowplow/tracker-core';
import { GaCookiesPlugin } from '../src';

describe('GA Cookies plugin', () => {
  let cookieJar: string;

  beforeAll(() => {
    cookieJar = '';
    jest.spyOn(document, 'cookie', 'set').mockImplementation((cookie) => {
      cookieJar += cookie;
    });
    jest.spyOn(document, 'cookie', 'get').mockImplementation(() => cookieJar);
  });

  afterEach(() => {
    cookieJar = '';
    jest.clearAllTimers();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('Returns values for Universal Analytics cookies by default', (done) => {
    const containerId = '1234';
    document.cookie = `_ga=1234; __utma=567; _ga_${containerId}=567;`;
    const core = trackerCore({
      corePlugins: [GaCookiesPlugin()],
      callback: (payloadBuilder) => {
        const { data } = payloadBuilder.getJson()[1].json;
        expect(data).toMatchSnapshot();
        done();
      },
    });

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });

  it('Does not add any additional GA4 context when GA4 cookies are not detected', (done) => {
    /* In v3, an empty Universal Analytics cookies context was added to the contexts array. */
    const containerId = '1234';
    const measurementId = `G-${containerId}`;
    document.cookie = ``;
    const core = trackerCore({
      corePlugins: [GaCookiesPlugin({ ga4: true, ga4MeasurementId: measurementId })],
      callback: (payloadBuilder) => {
        const { data } = payloadBuilder.getJson()[1].json;
        expect(data).toMatchSnapshot();
        done();
      },
    });

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });

  it('Returns values for GA4 cookies and Universal Analytics cookies', (done) => {
    const containerId = '1234';
    const measurementId = `G-${containerId}`;
    document.cookie = `_ga=1234; __utma=567; _ga_${containerId}=567;`;
    const core = trackerCore({
      corePlugins: [GaCookiesPlugin({ ga4: true, ga4MeasurementId: measurementId })],
      callback: (payloadBuilder) => {
        const { data } = payloadBuilder.getJson()[1].json;
        expect(data).toMatchSnapshot();
        done();
      },
    });

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });

  it('Returns values for GA4 cookies and not Universal Analytics cookies', (done) => {
    const containerId = '1234';
    const measurementId = `G-${containerId}`;
    document.cookie = `_ga=1234; __utma=567; _ga_${containerId}=567;`;
    const core = trackerCore({
      corePlugins: [GaCookiesPlugin({ ua: false, ga4: true, ga4MeasurementId: measurementId })],
      callback: (payloadBuilder) => {
        const { data } = payloadBuilder.getJson()[1].json;
        expect(data).toMatchSnapshot();
        done();
      },
    });

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });

  it('Correctly returns values for GA4 cookies for multiple measurement ids', (done) => {
    const containerIdFoo = '1234';
    const containerIdBar = '5678';
    const measurementIds = [`G-${containerIdFoo}`, `G-${containerIdBar}`];
    document.cookie = `_ga=1234; __utma=567; _ga_${containerIdFoo}=567; _ga_${containerIdBar}=789;`;
    const core = trackerCore({
      corePlugins: [GaCookiesPlugin({ ua: false, ga4: true, ga4MeasurementId: measurementIds })],
      callback: (payloadBuilder) => {
        const { data } = payloadBuilder.getJson()[1].json;
        expect(data).toMatchSnapshot();
        done();
      },
    });

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });

  it('Correctly returns values for GA4 cookies with cookie prefix', (done) => {
    const containerId = '1234';
    const measurementId = `G-${containerId}`;
    const cookiePrefix = 'test';
    document.cookie = `${cookiePrefix}_ga=1234; __utma=567; ${cookiePrefix}_ga_${containerId}=567;`;
    const core = trackerCore({
      corePlugins: [GaCookiesPlugin({ ua: false, ga4: true, ga4MeasurementId: measurementId, cookiePrefix })],
      callback: (payloadBuilder) => {
        const { data } = payloadBuilder.getJson()[1].json;
        expect(data).toMatchSnapshot();
        done();
      },
    });

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });

  it('Only returns values for GA4 user cookies without measurement id/s and GA4 enabled', (done) => {
    const containerId = '1234';
    document.cookie = `_ga=1234; __utma=567; _ga_${containerId}=567;`;
    const core = trackerCore({
      corePlugins: [GaCookiesPlugin({ ua: false, ga4: true })],
      callback: (payloadBuilder) => {
        const { data } = payloadBuilder.getJson()[1].json;
        expect(data).toMatchSnapshot();
        done();
      },
    });

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });

  it('Returns separate contexts for GA4 enabled setup with different cookie_prefix options', (done) => {
    const containerId = '1234';
    const cookiePrefix = 'test';
    /* gtag.js with the { cookie_prefix } option will create {cookie_prefix}_ga and {cookie_prefix}_ga_{containerId} cookies. */
    document.cookie = `_ga=1234; __utma=567; _ga_${containerId}=567; ${cookiePrefix}_ga=890;`;
    const core = trackerCore({
      corePlugins: [GaCookiesPlugin({ ua: false, ga4: true, cookiePrefix })],
      callback: (payloadBuilder) => {
        const { data } = payloadBuilder.getJson()[1].json;
        expect(data).toMatchSnapshot();
        done();
      },
    });

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });

  it('Returns separate contexts for GA4 enabled setup with different cookie_prefix options and measurement ids', (done) => {
    const containerId = '1234';
    const measurementId = `G-${containerId}`;
    const prefixContainerId = '567';
    const prefixMeasurementId = `G-${prefixContainerId}`;
    const cookiePrefix = 'test';
    /* gtag.js with the { cookie_prefix } option will create {cookie_prefix}_ga and {cookie_prefix}_ga_{containerId} cookies. */
    document.cookie = `_ga=1234; __utma=567; _ga_${containerId}=567; ${cookiePrefix}_ga=890; ${cookiePrefix}_ga_${prefixContainerId}=098`;
    const core = trackerCore({
      corePlugins: [
        GaCookiesPlugin({ ua: false, ga4: true, cookiePrefix, ga4MeasurementId: [measurementId, prefixMeasurementId] }),
      ],
      callback: (payloadBuilder) => {
        const { data } = payloadBuilder.getJson()[1].json;
        expect(data).toMatchSnapshot();
        done();
      },
    });

    core.track(buildLinkClick({ targetUrl: 'https://example.com' }));
  });
});
