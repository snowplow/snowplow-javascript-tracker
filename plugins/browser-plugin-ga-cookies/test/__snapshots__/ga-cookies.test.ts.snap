// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GA Cookies plugin Correctly returns values for GA4 cookies for multiple measurement ids 1`] = `
Array [
  Object {
    "data": Object {
      "_ga": "1234",
      "cookie_prefix": undefined,
      "session_cookies": Array [
        Object {
          "measurement_id": "G-1234",
          "session_cookie": "567",
        },
        Object {
          "measurement_id": "G-5678",
          "session_cookie": "789",
        },
      ],
    },
    "schema": "iglu:com.google.ga4/cookies/jsonschema/1-0-0",
  },
]
`;

exports[`GA Cookies plugin Correctly returns values for GA4 cookies with cookie prefix 1`] = `
Array [
  Object {
    "data": Object {
      "_ga": "1234",
      "cookie_prefix": "test",
      "session_cookies": Array [
        Object {
          "measurement_id": "G-1234",
          "session_cookie": "567",
        },
      ],
    },
    "schema": "iglu:com.google.ga4/cookies/jsonschema/1-0-0",
  },
]
`;

exports[`GA Cookies plugin Does not add any additional GA4 context when GA4 cookies are not detected 1`] = `
Array [
  Object {
    "data": Object {},
    "schema": "iglu:com.google.analytics/cookies/jsonschema/1-0-0",
  },
]
`;

exports[`GA Cookies plugin Only returns values for GA4 user cookies without measurement id/s and GA4 enabled 1`] = `
Array [
  Object {
    "data": Object {
      "_ga": "1234",
      "cookie_prefix": undefined,
      "session_cookies": undefined,
    },
    "schema": "iglu:com.google.ga4/cookies/jsonschema/1-0-0",
  },
]
`;

exports[`GA Cookies plugin Returns separate contexts for GA4 enabled setup with different cookie_prefix options 1`] = `
Array [
  Object {
    "data": Object {
      "_ga": "1234",
      "cookie_prefix": undefined,
      "session_cookies": undefined,
    },
    "schema": "iglu:com.google.ga4/cookies/jsonschema/1-0-0",
  },
  Object {
    "data": Object {
      "_ga": "890",
      "cookie_prefix": "test",
      "session_cookies": undefined,
    },
    "schema": "iglu:com.google.ga4/cookies/jsonschema/1-0-0",
  },
]
`;

exports[`GA Cookies plugin Returns separate contexts for GA4 enabled setup with different cookie_prefix options and measurement ids 1`] = `
Array [
  Object {
    "data": Object {
      "_ga": "1234",
      "cookie_prefix": undefined,
      "session_cookies": Array [
        Object {
          "measurement_id": "G-1234",
          "session_cookie": "567",
        },
      ],
    },
    "schema": "iglu:com.google.ga4/cookies/jsonschema/1-0-0",
  },
  Object {
    "data": Object {
      "_ga": "890",
      "cookie_prefix": "test",
      "session_cookies": Array [
        Object {
          "measurement_id": "G-567",
          "session_cookie": "098",
        },
      ],
    },
    "schema": "iglu:com.google.ga4/cookies/jsonschema/1-0-0",
  },
]
`;

exports[`GA Cookies plugin Returns values for GA4 cookies and Universal Analytics cookies 1`] = `
Array [
  Object {
    "data": Object {
      "__utma": "567",
      "_ga": "1234",
    },
    "schema": "iglu:com.google.analytics/cookies/jsonschema/1-0-0",
  },
  Object {
    "data": Object {
      "_ga": "1234",
      "cookie_prefix": undefined,
      "session_cookies": Array [
        Object {
          "measurement_id": "G-1234",
          "session_cookie": "567",
        },
      ],
    },
    "schema": "iglu:com.google.ga4/cookies/jsonschema/1-0-0",
  },
]
`;

exports[`GA Cookies plugin Returns values for GA4 cookies and not Universal Analytics cookies 1`] = `
Array [
  Object {
    "data": Object {
      "_ga": "1234",
      "cookie_prefix": undefined,
      "session_cookies": Array [
        Object {
          "measurement_id": "G-1234",
          "session_cookie": "567",
        },
      ],
    },
    "schema": "iglu:com.google.ga4/cookies/jsonschema/1-0-0",
  },
]
`;

exports[`GA Cookies plugin Returns values for Universal Analytics cookies by default 1`] = `
Array [
  Object {
    "data": Object {
      "__utma": "567",
      "_ga": "1234",
    },
    "schema": "iglu:com.google.analytics/cookies/jsonschema/1-0-0",
  },
]
`;
